src/service/blockchain-event-listener-service.js

```diff
@@ -437,15 +483,97 @@ class BlockchainEventListenerService {
+                const combinedData = eventsGroup.reduce((acc, event) => {
+                    try {
+                        const eventData = JSON.parse(event.data);
+                        return {
+                            ...acc,
+                            ...eventData,
+                            blockchainId: event.blockchainId,
+                        };
+                    } catch (error) {
+                        this.logger.error(`Error parsing event data: ${error}`);
+                        return acc;
+                    }
+                }, {});
 
-                const { epochsNumber } = await this.blockchainModuleManager.getAgreementData(
-                    event.blockchainId,
+                const {
+                    blockchainId,
+                    assetContract: contract,
+                    tokenId,
+                    keyword,
+                    hashFunctionId,
+                    state: assertionId,
+                    startTime,
+                    epochsNumber,
+                    epochLength,
+                    // TODO: Uncomment when these arguments are added to the ServiceAgreementV1Created event
+                    // scoreFunctionId,
+                    // proofWindowOffsetPerc,
+                } = combinedData;
+
+                let scoreFunctionId;
+                if (blockchainId.startsWith('otp')) {
+                    scoreFunctionId = 1;
+                } else if (blockchainId.startsWith('gnosis')) {
+                    scoreFunctionId = 2;
+                }
+
+                const tx = await this.blockchainModuleManager.getTransaction(
+                    blockchainId,
+                    eventsGroup[0].transactionHash,
+                );
+                const assetCreator = tx.from;
+                const minProofWindowOffsetPerc =
+                    await this.blockchainModuleManager.getMinProofWindowOffsetPerc(blockchainId);
+                const maxProofWindowOffsetPerc =
+                    await this.blockchainModuleManager.getMaxProofWindowOffsetPerc(blockchainId);
+
+                const proofWindowOffsetPerc =
+                    minProofWindowOffsetPerc +
+                    (await this.blockchainModuleManager.generatePseudorandomUint8(
+                        blockchainId,
+                        assetCreator,
+                        tx.blockNumber,
+                        tx.timestamp,
+                        maxProofWindowOffsetPerc - minProofWindowOffsetPerc + 1,
+                    ));
+
+                const agreementId = await this.serviceAgreementService.generateId(
+                    blockchainId,
+                    contract,
+                    tokenId,
+                    keyword,
+                    hashFunctionId,
+                );
+
+                await this.repositoryModuleManager.updateServiceAgreementRecord(
+                    blockchainId,
```

Comment:
Can we some identifier in service_agreement table in db so we know how it was created - during publish or after reacting on event, it will be easier for debugging later.

Reply:
New column perhaps?